cranelift/frontend/src/ssa.rs

```diff
@@ -281,64 +294,55 @@ impl SSABuilder {
-        // This function is split into two parts to appease the borrow checker.
-        // Part 1: With a mutable borrow of self, update the DataFlowGraph if necessary.
-        let data = &mut self.ssa_blocks[block];
-        let case = if data.sealed {
-            // Optimize the common case of one predecessor: no param needed. However, if following
-            // single-predecessor edges would go through a cycle ending up back here, then this
-            // optimization is unsafe, so fall back to adding a param anyway.
-            if data.predecessors.len() == 1 && !data.in_predecessor_cycle {
-                UseVarCases::SealedOnePredecessor(data.predecessors[0].block)
-            } else {
-                // Break potential cycles by eagerly adding an operandless param.
-                let val = func.dfg.append_block_param(block, ty);
-                UseVarCases::SealedMultiplePredecessors(val, block)
-            }
-        } else {
-            let val = func.dfg.append_block_param(block, ty);
-            data.undef_variables.push((var, val));
-            UseVarCases::Unsealed(val)
-        };
-
-        // Part 2: Prepare SSABuilder state for run_state_machine().
-        match case {
-            UseVarCases::SealedOnePredecessor(pred) => {
-                // Get the Value directly from the single predecessor.
-                self.calls.push(Call::FinishSealedOnePredecessor(block));
-                self.calls.push(Call::UseVar(pred));
-            }
-            UseVarCases::Unsealed(val) => {
-                // Define the operandless param added above to prevent lookup cycles.
-                self.def_var(var, val, block);
-
-                // Nothing more can be known at this point.
-                self.results.push(val);
-            }
-            UseVarCases::SealedMultiplePredecessors(val, block) => {
-                // Define the operandless param added above to prevent lookup cycles.
-                self.def_var(var, val, block);
+    fn use_var_nonlocal(&mut self, func: &mut Function, var: Variable, ty: Type, mut block: Block) {
+        let (val, from) = self.find_var(func, var, ty, block);
+        // Define the operandless param added above to prevent lookup cycles.
+        while block != from {
+            self.def_var(var, val, block);
+            let data = &self.ssa_blocks[block];
+            debug_assert!(data.sealed);
+            debug_assert_eq!(data.predecessors.len(), 1);
+            block = data.predecessors[0].block;
+        }
+    }
 
-                // Look up a use_var for each precessor.
-                self.begin_predecessors_lookup(val, block);
+    fn find_var(
+        &mut self,
+        func: &mut Function,
+        var: Variable,
+        ty: Type,
+        mut block: Block,
+    ) -> (Value, Block) {
+        let epoch = self.new_epoch();
+        while let Some(pred) = self.ssa_blocks[block].visit(epoch) {
```

Comment:
there's something subtle here about detecting/avoiding cycles vs. not visiting a block more than once. Not visiting more than once certainly avoids cycles; but do we need to do something else for them? Could you comment in which cases cycles may occur and why we are correct (insert the appropriate blockparams) in such cases?